### PR TITLE
refactor!: change many `U256` type to `U64`

### LINKED ISSUE #1553: Reduce the size for many fixed-size types, such as `gas_limit`, and so on.
### Proposal

Reduce the size for many fixed-size types.

For example, why `gas_limit` is defined as `U256` but used as `u64`? (ref: [evm::StackSubstateMetadata::new(gas_limit: u64, config: &'config Config)](https://docs.rs/evm/0.41.0/evm/executor/stack/struct.StackSubstateMetadata.html#method.new))

It could lead to panic when [convert between `U256` and `u64`](https://github.com/paritytech/parity-common/blob/b4e1516a95c7e7c0bde66d03dbea6f388e7c5fec/uint/src/uint.rs#L644-L650), and the conversion is not equivalent (64 bits vs 256 bits).

The same situation for `number`, `nonce` and some other fields.

### Reasons

Here, I use `gas_limit` as the example to explain:

- `gas_limit` in `Header`:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/protocol/src/types/block.rs#L84
- `gas_limit` in `Transaction`:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/protocol/src/types/transaction.rs#L188
- `gas_limit` in `Consensus`:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/protocol/src/types/primitive.rs#L400
- `gas_limit` in `Executor`:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/core/executor/src/lib.rs#L71
- Constants of `gas_limit`:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/protocol/src/types/block.rs#L16
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/protocol/src/types/transaction.rs#L21
- `gas_limit` in pre-compiled contracts:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/core/executor/src/precompiles/mod.rs#L72
- `gas_limit` in system contracts:
  https://github.com/axonweb3/axon/blob/0fb4d7f02dddb126eb1e0ca9e2ac2121cc29fac8/core/executor/src/system_contract/utils.rs#L30-L31

### References

- [Chapter 4.3 in Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
- [Block definition in Go Ethereum](https://github.com/ethereum/go-ethereum/blob/f265cc24b4b17b55b468b315303a3a599fc01d81/core/types/block.go#L73-L77)
- [Block definition in Lighthouse (A Ethereum Consensus Client in Rust)](https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/consensus/types/src/execution_payload.rs#L65-L74) 
- [Block definition in Helios (A Ethereum light client written in Rust)](https://github.com/a16z/helios/blob/8a4323a159f5fa79db865e536666a568837f4ad1/common/src/types.rs#L9-L14)
- [`gas_limit` in Cosmos and Hermes (A Rust implementation of an Inter-Blockchain Communication (IBC) Relayer)](https://github.com/cosmos/ibc-proto-rs/blob/1075017fbcb110b5ca2686dad4dcedc439b1ece1/src/prost/cosmos.tx.v1beta1.rs#L238)

### LINKED ISSUE #1584: Duplicated calculation in `eth_estimateGas`.
### Description

When call `eth_estimateGas` to estimate gas used, the gas calculation for `create-tx`, `call-tx` and `tx.data` are incorrect and duplicated.

https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/core/executor/src/lib.rs#L81-L85

https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/core/executor/src/lib.rs#L100

---

### For Example

- Do simplest transfer transaction.
  | Data Size[^1] | Estimated in Axon | Actual in Axon  | Estimated in Geth     | Actual in Geth      |
  |--------------:|------------------:|----------------:|----------------------:|--------------------:|
  |             0 |             42000 |           21000 |                 21000 |               21000 |
  |             1 |             42084 |           21016 |                 21016 |               21016 |
  |             2 |             42168 |           21032 |                 21032 |               21032 |

  [^1]: Set the `transaction.data` to be `N` non-zero bytes.

---

### Appendix: More Unit Tests are Required

- Call `eth_estimateGas` for a transaction whose gas is set to a value that greater than [`u64::MAX`], will cause Axon panic.

  https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/core/api/src/adapter.rs#L210

  There is already an in-process PR.

  I just record it here, and request a unit test for it, after that PR merged.

[`u64::MAX`]: https://doc.rust-lang.org/std/primitive.u64.html#associatedconstant.MAX

### LINKED ISSUE #1586: bug: bypass restrictions through p2p message, even cause panic
### Description

Submit transactions from different components leads different results.

Even some p2p messages could let all nodes panic.

p.s. All server limits in [`config.toml`] are ignored, since they are not consensus parameters.

### Examples

- [x] A transaction whose gas limit is less than [`MIN_TRANSACTION_GAS_LIMIT`].

  JSON-RPC Post: Failed with ["Gas limit is less than 21000"](https://github.com/axonweb3/axon/blob/main/core/api/src/jsonrpc/impl/web3.rs#L208).
  P2P Broadcast: Succeed.

- [x] A transaction whose gas limit is greater than [`u64::MAX`].

  JSON-RPC Post: Failed with "Gas limit is too large".
  P2P Broadcast: [Panic](https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/core/mempool/src/adapter/mod.rs#L265).

- [x] Two transactions whose used gas is greater than a half of [`u64::MAX`].

  JSON-RPC Post: Failed with "Gas limit is too large".
  P2P Broadcast: [Panic](https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/core/executor/src/lib.rs#L156).

I just list some of the unexpected cases, I think these are not all of them.

Please check the code again, carefully and imaginatively.

### References

- #1582

[`config.toml`]: https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/devtools/chain/config.toml
[`MIN_TRANSACTION_GAS_LIMIT`]: https://github.com/axonweb3/axon/blob/f692b865bda85cbd0c8f028f8aca8ae0d793465b/protocol/src/types/transaction.rs#L21
[`u64::MAX`]: https://doc.rust-lang.org/std/primitive.u64.html#associatedconstant.MAX

### PR DESCRIPTION
**Pull request by KaoImin**
## What this PR does / why we need it?

This PR is substitute for #1539

Some mainly changes:
* Change the type of `nonce` from `U256` to `U64` according to the [EIP-2681](https://eips.ethereum.org/EIPS/eip-2681) limit the account nonce to be between `0` and `2^64-1`.
 * Change the type of `gas_limit` from `U256` to `U64`. According to the current gas cost of the most complex ethereum transaction is on the order of million, `U64` is enough.
* Change the type of `chain_id` from `U256` to `U64` according to  [metamask limit](https://gist.github.com/rekmarks/a47bd5f2525936c4b8eee31a16345553). The [`ChainId` opcode](https://eips.ethereum.org/EIPS/eip-1344) returns a 256-bit value, so it should not larger than `U256::MAX / 2 - 35`.This issue does not lead to consensus in Ethereum community, however,  I think limit the max chain ID to `4503599627370476` is enough currently. Many temporary L2/L3 need a unique chain ID is the demand of variable-lenght chain ID in foreseeable future. But we do not have the demand now. If the day comes, we can change the chain ID type to `U256` even without hardfork.
* Set the [`Default Max Price`](https://github.com/ethereum/go-ethereum/blob/be65b47/eth/gasprice/gasprice.go#L38) as `500` Gwei that is same as go-ethereum. Meanwhile, change the type of `gas_price` from `U256` to `U64`.

### What is the impact of this PR?
Breaking change

**PR relation**:
- Ref #1539
- Fix #1586
- test code updated: https://github.com/axonweb3/axon-test/compare/be8b502423fa41d4656ec2369d3ee5a729f56326...df72c7e1660537b445a349624b9b3b1584036393

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [x] Web3 Compatible Tests
- [x] OpenZeppelin tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [x] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

**Comment by chaoticlonghair** [comment]
References:
- [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
  - 4.2. The Transaction
    - `nonce`: a scalar value
    - `gasPrice`: a scalar value
    - `gasLimit`: a scalar value
    - `value`: a scalar value
    - `chainId`: unknown
  - 4.3. The Block
    - `difficulty`: a scalar value
    - `number`: a scalar value
    - `gasLimit`: a scalar value
    - `gasUsed`: a scalar value
    - `timestamp`: a scalar value
    - `nonce`: a 64-bit value
- Go Etherem:
  - [Transaction](https://github.com/ethereum/go-ethereum/blob/333dd956bfdf1d5086d38cceedbba25a366fb6ac/core/types/transaction.go#L74)
    ```go
    chainID() *big.Int
    gas() uint64
    gasPrice() *big.Int
    gasTipCap() *big.Int
    gasFeeCap() *big.Int
    value() *big.Int
    nonce() uint64
    ```
  - [Block](https://github.com/ethereum/go-ethereum/blob/333dd956bfdf1d5086d38cceedbba25a366fb6ac/core/types/block.go#L73-L77)
    ```go
    Difficulty  *big.Int       `json:"difficulty"       gencodec:"required"`
    Number      *big.Int       `json:"number"           gencodec:"required"`
    GasLimit    uint64         `json:"gasLimit"         gencodec:"required"`
    GasUsed     uint64         `json:"gasUsed"          gencodec:"required"`
    Time        uint64         `json:"timestamp"        gencodec:"required"`
    Extra       []byte         `json:"extraData"        gencodec:"required"`
    MixDigest   common.Hash    `json:"mixHash"`
    Nonce       BlockNonce     `json:"nonce"`
    ```

**Comment by KaoImin** [comment]
> References:
> 
> [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
> 
> 4.2. The Transaction
> 
> *   `nonce`: a scalar value
> *   `gasPrice`: a scalar value
> *   `gasLimit`: a scalar value
> *   `value`: a scalar value
> *   `chainId`: unknown
> 
> 4.3. The Block
> 
> *   `difficulty`: a scalar value
> *   `number`: a scalar value
> *   `gasLimit`: a scalar value
> *   `gasUsed`: a scalar value
> *   `timestamp`: a scalar value
> *   `nonce`: a 64-bit value
> 
> Go Etherem:
> 
> *   [Transaction](https://github.com/ethereum/go-ethereum/blob/333dd956bfdf1d5086d38cceedbba25a366fb6ac/core/types/transaction.go#L74)
> *   [Block](https://github.com/ethereum/go-ethereum/blob/333dd956bfdf1d5086d38cceedbba25a366fb6ac/core/types/block.go#L73-L77)
> 
> ```go
> Difficulty  *big.Int       `json:"difficulty"       gencodec:"required"`
> Number      *big.Int       `json:"number"           gencodec:"required"`
> GasLimit    uint64         `json:"gasLimit"         gencodec:"required"`
> GasUsed     uint64         `json:"gasUsed"          gencodec:"required"`
> Time        uint64         `json:"timestamp"        gencodec:"required"`
> Extra       []byte         `json:"extraData"        gencodec:"required"`
> MixDigest   common.Hash    `json:"mixHash"`
> Nonce       BlockNonce     `json:"nonce"`
> ```
> 
> ```go
> chainID() *big.Int
> gas() uint64
> gasPrice() *big.Int
> gasTipCap() *big.Int
> gasFeeCap() *big.Int
> value() *big.Int
> nonce() uint64
> ```

According to the PR description, some types such as `nonce`, `gasPrice`, `gasLimit`, `gasUsed` have a specific range limitation, it is a waste of performance to use a variable-length `BigUint`. The Ethereum `timestamp` is accurate to the second, so `u64` is enough. The only one that using variable-length scalar is `chainId`. The explanation of using `U64` is written in the description.

**Review by driftluo — DISMISSED** [review]

**Comment by KaoImin** [comment]
Wait for @sunchengzhu upgrade `Web3 Compatible Tests` workflow. cc @Flouse

**Comment by Flouse** [comment]
/run-ci

run OpenZeppelin tests as well

File: protocol/src/constants/configs.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+pub const MAX_GAS_PRICE: U64 = U64([500 * GWEI]);
+pub const MIN_TRANSACTION_GAS_LIMIT: u64 = 21_000;
+/// The mempool refresh timeout is 50 milliseconds.
+pub const MEMPOOL_REFRESH_TIMEOUT: u64 = 50;
+pub const MAX_BLOCK_GAS_LIMIT: u64 = 30_000_000;
+// MAX_FEE_HISTORY is the maximum number of blocks that can be retrieved for a
+// fee history request. Between 1 and 1024 blocks can be requested in a single
+// query. reference: https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_feehistory/
+pub const MAX_FEE_HISTORY: u64 = 1024;
+pub const MAX_RPC_GAS_CAP: u64 = 50_000_000;
```
Comment (Flouse) [inline protocol/src/constants/configs.rs]: MAX_RPC_GAS_CAP is not used, right?

Comment (KaoImin) [reply protocol/src/constants/configs.rs]: Yes

File: protocol/src/types/primitive.rs [hunk 2]
```diff
@@ -174,10 +174,6 @@ impl<'de> Deserialize<'de> for Hex {
 }
 
 /// A 256 bits bytes used for sensitive data, such as private keys.
-///
-/// It's implemented a `Drop` handler which erase its memory when it dropped.
-/// This ensures that sensitive data is securely erased from memory when it is
-/// no longer needed.
 pub type Key256Bits = Zeroizing<[u8; 32]>;
```
Comment (Flouse) [inline protocol/src/types/primitive.rs]: What is the reason for removing this comment?

Comment (KaoImin) [reply protocol/src/types/primitive.rs]: This is not my operation.

Comment (Flouse) [reply protocol/src/types/primitive.rs]: What do you mean it is not your operation?
It was deleted in your commit
https://github.com/axonweb3/axon/commit/96ba83c5a907481765be33067f89441f225d7c4f#diff-9e85d32a5e7a6dc441136d66a31b595ef11780890af67a35b58fd72d69a0d950L180

Comment (KaoImin) [reply protocol/src/types/primitive.rs]: I got it, it is removed by `cargo fmt`.

**Review by Flouse — DISMISSED** [review]

**Review by driftluo — DISMISSED** [review]

File: core/executor/src/tests/mod.rs [hunk 3]
```diff
@@ -127,20 +127,6 @@ async fn test_simplestorage() {
     let precompiles = build_precompile_set();
 
     // pragma solidity ^0.4.24;
```
Comment (Flouse) [inline core/executor/src/tests/mod.rs]:
```suggestion
    // pragma solidity ^0.4.24;
    //
    // contract SimpleStorage {
    //     uint storedData;
    //
    //     function set(uint x) public {
    //         storedData = x;
    //     }
    //
    //     function get() view public returns (uint) {
    //         return storedData;
    //     }
    // }
    //
    // simplestorage_create_code created from above solidity
```

**Review by Flouse — APPROVED** [review]

**Pull request merged**